File: OOPs/Basics/Static_KW.cpp
#include<iostream>
using namespace std;

class RoyalStudent {
    string name;
    enum TYPEOFSTUDENTS{GENERAL=400000, CLUB=700000, ONE2ONE=1500000};
    int fees;

    public:
        static int totalStudents;
        RoyalStudent(string name, int type) {
            this->name = name;
            //here we will set the fees according to the type of student
            if(type == 1) {
                fees = GENERAL;
            } else if(type == 2) {
                fees = CLUB;
            } else if(type == 3) {
                fees = ONE2ONE;
            }
            totalStudents++;
        }

        void show() {
            cout << "Name: " << name << endl;
            cout << "Total Students: " << totalStudents << endl;
        }

        static int getGeneralFees() {
            return GENERAL;
        }

        static int getClubFees() {
            return CLUB;
        }

        static int getOne2OneFees() {
            return ONE2ONE;
        }
};

//we have to declare the static variable outside the class
int RoyalStudent::totalStudents = 0;

int getGeneralFees() {
    return 0;
}

int main() {

    RoyalStudent r1("Kaushal Ramanuj", 2);
    RoyalStudent r2("Himanshu Sorathiya", 1);
    RoyalStudent r3("Jay Vaghasiya", 3);

    r1.show();
    // r2.show();
    // r3.show();

    cout << "General Fees: " << RoyalStudent::getGeneralFees() << endl;
    cout << "Club Fees: " << RoyalStudent::getClubFees() << endl;
    cout << "One2One Fees: " << RoyalStudent::getOne2OneFees() << endl;

    cout << "Total Students: " << RoyalStudent::totalStudents << endl;
    cout << "Gen Fees (Without Parents): " << getGeneralFees() << endl;

    return 0;
}

/*
> Static Keyword in cpp is used to give a special property to a variable or a function.
> It allows to use the variable or function without creating an object of the class.
> Things which are declared as static are stored in the static memory, means they are stored in the memory which is allocated at the time of compilation.

Taskks:
1. Static Member Variables:

   Design a C++ program to create a class called 'MyClass' with a static integer variable named 'staticVar'. Create two instances of this class, 'obj1' and 'obj2'. Set the 'staticVar' of 'obj1' to 42 and display the values of 'staticVar' for both objects. Explain the behavior of static variables in different instances of the same class.

2. Static Member Functions:

   Create a C++ program that defines a class 'MyClass' with a static member function called 'staticFunction'. Use this function to return the value 42 and print the result. Discuss the usage and advantages of static member functions in C++.

3. Static Local Variables:

   Write a C++ function called 'demoStaticLocalVariable' that contains a static local integer variable named 'count'. Call this function multiple times and observe the changes in the value of 'count'. Explain the significance of static local variables and how they maintain their state across function calls.

4. Static Member Initialization:

   Create a C++ class named 'MyClass' with two static integer variables, 'staticVar1' and 'staticVar2'. Initialize 'staticVar1' to 42 and 'staticVar2' to 100 using different methods. Write a program to display the values of these static variables and compare the initialization techniques.

5. Static Class Variables:

   Implement a C++ program to simulate a simple bank account system using a class called 'BankAccount'. Inside this class, declare a static integer variable 'totalAccounts' to keep track of the total number of bank accounts. Create multiple instances of 'BankAccount', open new accounts, and close existing accounts. Display the current value of 'totalAccounts' after each operation. Discuss the role of static class variables in maintaining shared information among objects.

These problem statements will guide you in exploring various aspects of the 'static' keyword in C++ and its applications in different scenarios.

*/

File: OOPs/Basics/Static_KW.java
import java.util.Scanner;

class Student {
    int rollNo;
    String name;
    static String college;

    static {
        System.out.println("Static Block");
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter college: ");
        college = sc.next();
        sc.close();
    }

    Student(int rollNo, String name) {
        System.out.println("Constructor");
        this.rollNo = rollNo;
        this.name = name;
    }

    static void display(Student s) {
        // System.out.println("RollNo: " + rollNo + "\tName: " + name + "\tCollege Name: " + college);
        System.out.println(s.rollNo + "\t\t" + s.name + "\t\t" + college);
    }
}

//

public class Static_KW {
    public static void main(String[] args) {
        Student s[] = new Student[3];
        
        s[0]= new Student(1, "Kaushal");
        s[1]= new Student(2, "Shally");
        Student.college = "LDRP";
        s[2]= new Student(3, "Dhruvi");
        System.out.println("RollNo\t\tName\t\tCollege");
        Student.display(s[0]);
        Student.display(s[1]);
        Student.display(s[2]);

    }
}

/*
static keyword in java is used for memory management mainly.
static - variable, method, block

any thing which is declared as static in java will be allocated with memory only once. then it will share the same memory to all of it's instances. 

static things stored at static memory location

---------------------------------------------------------

Static is used to refer the common property of all objects.

Anything that will be declared as static will be common for all the objects of that class.

Task:

Problem Statement: Car Fleet Management

You are tasked with developing a car fleet management system. The system needs to keep track of various pieces of information for each car in the fleet, such as the make, model, year, price, and whether the car is currently in use. Additionally, the system should maintain a count of the total number of cars in the fleet.

Design and implement a `Car` class to represent individual cars in the fleet. The `Car` class should have the following features:

- Attributes:
  - `make` (String): The make or brand of the car.
  - `model` (String): The model of the car.
  - `year` (int): The manufacturing year of the car.
  - `price` (double): The price of the car.
  - `isRunning` (boolean): A flag indicating whether the car is currently running or not.

- Static Attribute:
  - `numberOfCars` (int): A static variable to keep track of the total number of cars in the fleet.

- Methods:
  - A constructor to initialize the car's attributes.
  - `start()` and `stop()` methods to change the `isRunning` status.
  - `displayCarInfo()` method to display car information.
  - A `getNumberOfCars()` static method to retrieve the total number of cars in the fleet.

You need to implement the `Car` class and use it to create, manage, and display information for multiple cars in the fleet. Ensure that the `numberOfCars` variable is properly incremented when new cars are added to the fleet and that it can be accessed without creating an instance of the `Car` class.




*/